.travis.yml

```diff
@@ -63,9 +60,9 @@ matrix:
     - python: 3.5
 
 before_install:
-  - wget -O geckodriver.tar.gz https://github.com/mozilla/geckodriver/releases/download/v0.12.0/geckodriver-v0.12.0-linux64.tar.gz
-  - gunzip -c geckodriver.tar.gz | tar xopf -
-  - chmod +x geckodriver && sudo mv geckodriver /usr/local/bin
+  - wget -O /tmp/geckodriver.tar.gz https://github.com/mozilla/geckodriver/releases/download/v0.12.0/geckodriver-v0.12.0-linux64.tar.gz
+  - mkdir $HOME/geckodriver && gunzip -c /tmp/geckodriver.tar.gz | tar xopf - -C $HOME/geckodriver
```

Comment:
This patch was more about switching to container based. I'm sure there are many other optimisations we can make independently. If you feel strongly I'll apply your suggestions in this pull, but as separate commits.